packages/flutter/lib/src/cupertino/text_selection.dart

```diff
@@ -270,49 +277,83 @@ class _CupertinoTextSelectionControls extends TextSelectionControls {
-        ? _ArrowDirection.down
-        : _ArrowDirection.up;
-
-    final TextSelectionPoint startTextSelectionPoint = endpoints[0];
-    final TextSelectionPoint endTextSelectionPoint = (endpoints.length > 1)
-        ? endpoints[1]
-        : null;
-    final double x = (endTextSelectionPoint == null)
-        ? startTextSelectionPoint.point.dx
-        : (startTextSelectionPoint.point.dx + endTextSelectionPoint.point.dx) / 2.0;
-    final double y = (direction == _ArrowDirection.up)
-        ? startTextSelectionPoint.point.dy + globalEditableRegion.height + _kToolbarHeight
-        : startTextSelectionPoint.point.dy - globalEditableRegion.height;
-    final Offset preciseMidpoint = Offset(x, y);
-
-    return ConstrainedBox(
-      constraints: BoxConstraints.tight(globalEditableRegion.size),
-      child: CustomSingleChildLayout(
-        delegate: _TextSelectionToolbarLayout(
-          MediaQuery.of(context).size,
-          globalEditableRegion,
-          preciseMidpoint,
-        ),
-        child: _TextSelectionToolbar(
-          handleCut: canCut(delegate) ? () => handleCut(delegate) : null,
-          handleCopy: canCopy(delegate) ? () => handleCopy(delegate) : null,
-          handlePaste: canPaste(delegate) ? () => handlePaste(delegate) : null,
-          handleSelectAll: canSelectAll(delegate) ? () => handleSelectAll(delegate) : null,
-          arrowDirection: direction,
+    // The y-coordinate has to be calculated instead of directly quoting postion.dy,
+    // since the caller (TextSelectionOverlay._buildToolbar) does not know whether
+    // the toolbar is going to be facing up or down.
+    final double localBarTopY = isArrowPointingDown
+      ? endpoints.first.point.dy - textLineHeight - _kToolbarContentDistance - _kToolbarHeight
+      : endpoints.last.point.dy + _kToolbarContentDistance;
+
+    final List<Widget> items = <Widget>[];
+    final Widget onePhysicalPixelVerticalDivider =
+    SizedBox(width: 1.0 / MediaQuery.of(context).devicePixelRatio);
+    final CupertinoLocalizations localizations = CupertinoLocalizations.of(context);
+    final EdgeInsets arrowPadding = isArrowPointingDown
+      ? EdgeInsets.only(bottom: _kToolbarArrowSize.height)
+      : EdgeInsets.only(top: _kToolbarArrowSize.height);
+
+    void addToolbarButtonIfNeeded(
+      String text,
+      bool Function(TextSelectionDelegate) predicate,
+      void Function(TextSelectionDelegate) onPressed
+    ) {
+      if (onPressed == null) {
+        return;
+      }
+
+      if (items.isNotEmpty) {
+        items.add(onePhysicalPixelVerticalDivider);
+      }
+
+      items.add(CupertinoButton(
+        child: Text(text, style: _kToolbarButtonFontStyle),
+        color: _kToolbarBackgroundColor,
+        minSize: _kToolbarHeight,
+        padding: _kToolbarButtonPadding.add(arrowPadding),
+        borderRadius: null,
+        pressedOpacity: 0.7,
+        onPressed: predicate(delegate) ? () => onPressed(delegate) : null
```

Comment:
Not sure why when the predicate evaluates to `false` we're still showing the button. I've never seen any disabled text selection toolbar button on iOS.